docs/proposals/20191016-e2e-test-framework.md

```diff
@@ -0,0 +1,199 @@
+
+The exact interface that a provider must satisfy is to be determined by what the tests demand. The plan for implementation at a high level is to build out reusable but pluggable testing components. For example, the framework will need a way to start a management cluster. We will use KIND initially and abstract the actual details away and expose an interface. If you want to use a different style of management cluster, say, a cloud-formation based cluster, then you will have to wrap up the cloud-formation code in some object and implement the testing framework’s exposed interface for the management cluster and use your object where the sample code uses the KIND-based management cluster struct. I expect but cannot guarantee that there will be interfaces for each provider but exact details are to be determined as the implementation gets more fleshed out.
+
+#### Behaviors to test
+
+* There is a Kubernetes cluster with one node that passes a healthcheck after creating a properly configured Cluster, InfraCluster, Machine, InfraMachine and BootstrapConfig.
+* Creating the resources necessary for three control planes will create a three node cluster.
+* Deleting a cluster deletes all resources associated with that cluster including Machines, BootstrapConfigs and generated secrets.
+* Creating a cluster with one control plane and one worker node will result in a cluster with two nodes.
+* The version fields in Machines are respected within the bounds of the Kubernetes skew policy.
+* Creating a control plane machine and a MachineDeployment with two replicas will create a three node cluster with one control plane node and two worker nodes.
+* Modifying a replica count on a MachineDeployment will modify the number of worker nodes the cluster has.
+* Optionally, Machines report failures when their underlying InfraMachines report failures.
+* Workload Clusters created pass Kubernetes Conformance
+
+### Risks and Mitigations
+
+These tests have two main risks: False positives/negatives and long-term technical debt. 
+
+We want these E2E tests to be useful for providers to use in a test-infra prow job. Depending on the providers’ choices, these tests will be able to be run as a regular test suite. Therefore they can run as pre jobs, post jobs, periodics or any other type of job that Prow supports. *Ideally these tests will replace, at least partially, the need for manual testing*. If the tests are too brittle, development velocity could take a hit, and if they’re too lenient, bugs are more likely to make it into production.
+
+To keep these tests healthy and useful, sig-cluster-lifecycle will need clear ownership on this common testing infrastructure. Keeping these tests happy and healthy across providers will be key to success, not just now, but going forward as well
```

Comment:
Should we scope this down from `sig-cluster-lifecycle` to just the `cluster-api` project?